### PR TITLE
update convict from 0.0.6 to 0.4.0

### PR DESCRIPTION
**Pull request by seanmonstar**
anti bit rot

### CODE REVIEW

**Comment by seanmonstar** [comment]
well dangit, i didnt realize 0.4.0 didn't parse JSON out of ENV variables. So I've submitted https://github.com/lloyd/node-convict/pull/19, and then I'll update this to the next version (0.4.1?).

**Comment by lloyd** [comment]
`node-convict` transferred to mozilla, v0.4.1 tagged and published to npm.  :heart:

**Comment by shane-tomlinson** [comment]
@seanmonstar - bump! @lloyd published a new node-convict.

**Comment by seanmonstar** [comment]
I moved all the kpi properties under a kpi namespace, so that might require config files to be adjust in stage/prod...

File: lib/configuration.js [hunk 1]
```diff
@@ -42,267 +42,382 @@ var conf = module.exports = convict({
...
   },
   certificate_validity_ms: {
     doc: "For how long shall certificates issued by BrowserID be valid?",
-    format: 'integer = 86400000'
+    format: 'duration',
+    default: 86400000
   },
   max_compute_processes: {
     doc: "How many computation processes will be spun.  Default is good, based on the number of CPU cores on the machine.",
-    format: 'union { number{1, 256}; null; } = null',
```
Comment (shane-tomlinson) [inline lib/configuration.js:218]: What was this trying to say?

File: lib/configuration.js [hunk 2]
```diff
@@ -341,10 +456,6 @@ if (process.env.CONFIG_FILES) {
   });
 }
 
-// allow supported langauges to be specified in the env as a CSV string
```
Comment (shane-tomlinson) [inline lib/configuration.js:344]: I totally dig these tidy ups.

**Comment by shane-tomlinson** [comment]
@seanmonstar - whenever installing on an ephemeral instance, I keep seeing:

```
npm http 404 http://127.0.0.1:38062/cjson/0.0.6
npm ERR! registry error parsing json
npm ERR! Error: failed to fetch from registry: cjson/0.0.6
npm ERR!     at RegClient.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/npm-registry-client/lib/get.js:134:18)
npm ERR!     at cb (/usr/local/lib/node_modules/npm/node_modules/npm-registry-client/lib/request.js:27:9)
npm ERR!     at RegClient.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/npm-registry-client/lib/request.js:148:10)
npm ERR!     at cb (/usr/local/lib/node_modules/npm/node_modules/npm-registry-client/lib/request.js:158:9)
npm ERR!     at RegClient.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/npm-registry-client/lib/request.js:243:16)
npm ERR!     at Request.self.callback (/usr/local/lib/node_modules/npm/node_modules/request/index.js:148:22)
npm ERR!     at Request.EventEmitter.emit (events.js:98:17)
npm ERR!     at Request.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/request/index.js:891:14)
npm ERR!     at Request.EventEmitter.emit (events.js:117:20)
npm ERR!     at IncomingMessage.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/request/index.js:842:12)
npm ERR! If you need help, you may report this log at:
npm ERR!     <http://github.com/isaacs/npm/issues>
npm ERR! or email it to:
npm ERR!     <npm-@googlegroups.com>

npm ERR! System Linux 3.4.57-48.42.amzn1.x86_64
```

After blowing away my `/node_modules` subdir locally, I see:

```
LOCKDOWN ERRORS:
    package version cjson@0.0.6 not in lockdown.json!
```

**Comment by shane-tomlinson** [comment]
@seanmonstar - with this diff, I can install on aws, but then receive a bunch of errors:

``` diff
commit 20812a6d077834ea40d7ac38b9d4b2ec40025931
Author: Shane Tomlinson <stomlinson@mozilla.com>
Date:   Mon Oct 14 18:58:27 2013 +0100

    Update lockdown with cjson@0.0.6

diff --git a/lockdown.json b/lockdown.json
index 48f8396..bb192e5 100644
--- a/lockdown.json
+++ b/lockdown.json
@@ -2,9 +2,6 @@
   "JSONSelect": {
     "0.4.0": "a08edcc67eb3fcbe99ed630855344a0cf282bb8d"
   },
-  "JSV": {
-    "3.5.0": "f017ec75629e8c89b8359dc79a2ceee92ef9471e"
-  },
   "addressparser": {
     "0.1.3": "9e9ab43d257e1ae784e1df5f580c9f5240f58874"
   },
@@ -58,6 +55,7 @@
     "1.0.2": "55384d6afcf8c6b9dd483e8347646a790e4545e7"
   },
   "cjson": {
+    "0.0.6": "f53a8664cd5f195c0bed06235cd4a38df88b4208",
     "0.2.1": "73cd8aad65d9e1505f9af1744d3b79c1527682a5"
   },
   "cli": {
@@ -277,9 +275,6 @@
   "mailcomposer": {
     "0.1.33": "69f6e924c55804f17c1bfea6c3f2f51049650820"
   },
-  "mersenne": {
-    "0.0.3": "c39a3d45fee6091189ccd329107312ea8fe14d8a"
-  },
   "mime": {
     "1.2.11": "58203eed86e3a5ef17aed2b7d9ebd47f0a60dd10",
     "1.2.6": "b1f86c768c025fa87b48075f1709f28aeaf20365",
@@ -325,8 +320,7 @@
     "0.3.21": "d9637a1d585f36fd30362c3036ce0692656fe771"
   },
   "nomnom": {
-    "1.5.2": "f4345448a853cfbd5c0d26320f2477ab0526fe2f",
-    "1.6.1": "bfed4506642d81278738e891c557e80694c1e0c9"
+    "1.5.2": "f4345448a853cfbd5c0d26320f2477ab0526fe2f"
   },
   "normalize-package-data": {
     "0.1.7": "f216a318a868214892e89ad80a944cddc1a79615",
@@ -349,9 +343,6 @@
     "0.3.7": "c90941ad59e4273328923074d2cf2e7cbc6ec0d9",
     "0.5.0": "d9c60da4c34811418d183390623f8046f134a2d4"
   },
-  "orderly": {
-    "1.0.1": "fb4e983db7900832373a7b0c903669c4dae78fb6"
-  },
   "osenv": {
     "0.0.3": "cd6ad8ddb290915ad9e22765576025d411f29cb6"
   },
```

These are the errors I see:

``` bash
remote: /home/app/code/node_modules/convict/lib/convict.js:377
remote:         throw new Error(errBuf);
remote:               ^
remote: Error: max_compute_processes: must be an integer
remote:     at Object.rv.validate (/home/app/code/node_modules/convict/lib/convict.js:377:15)
remote:     at Object.<anonymous> (/home/app/code/lib/configuration.js:498:6)
remote:     at Module._compile (module.js:456:26)
remote:     at Object.Module._extensions..js (module.js:474:10)
remote:     at Module.load (module.js:356:32)
remote:     at Function.Module._load (module.js:312:12)
remote:     at Module.require (module.js:364:17)
remote:     at require (module.js:380:17)
remote:     at Object.<anonymous> (/home/app/code/scripts/compress:6:10)
remote:     at Module._compile (module.js:456:26)
remote: ERROR: while postdeploy hook:
remote:
remote: Error: Command failed:
remote: /home/app/code/node_modules/convict/lib/convict.js:377
remote:         throw new Error(errBuf);
remote:               ^
remote: Error: max_compute_processes: must be an integer
remote:     at Object.rv.validate (/home/app/code/node_modules/convict/lib/convict.js:377:15)
remote:     at Object.<anonymous> (/home/app/code/lib/configuration.js:498:6)
remote:     at Module._compile (module.js:456:26)
remote:     at Object.Module._extensions..js (module.js:474:10)
remote:     at Module.load (module.js:356:32)
remote:     at Function.Module._load (module.js:312:12)
remote:     at Module.require (module.js:364:17)
remote:     at require (module.js:380:17)
remote:     at Object.<anonymous> (/home/app/code/scripts/compress:6:10)
remote:     at Module._compile (module.js:456:26)
remote:
To app@54.216.240.120:git
   f1ac042..20812a6  HEAD -> master
➜  ~PERSONA_ROOT git:(convict-0.4.0) ✗
```

It looks like `aws.json` or one of the other config files could use some love?

File: lib/configuration.js (hunk 1 above)
Comment (seanmonstar) [inline lib/configuration.js:218]: my guess is `1-256 || null`

**Comment by seanmonstar** [comment]
@shane-tomlinson nice catch! fixxored it up.

File: lib/configuration.js (hunk 1 above)
Comment (seanmonstar) [inline lib/configuration.js:218]: only using integers, and setting the default to 0, has the same result as using null, thanks to this truthy check: https://github.com/lloyd/node-compute-cluster/blob/master/lib/compute-cluster.js#L18

So, same effect, less union brain-boiling madness.

**Comment by shane-tomlinson** [comment]
@seanmonstar - I am now getting:

```
remote: >> generating ver.txt
remote: >> generating production resources
remote:
remote: /home/app/code/node_modules/cjson/lib/cjson.js:182
remote:         throw new Error('JSON.parse error - "' + e.message + '"\nFile: "' + pa
remote:               ^
remote: Error: JSON.parse error - "Unexpected token '"
remote: File: "/home/app/code/config/aws.json"
remote:
remote:     at Object.load (/home/app/code/node_modules/cjson/lib/cjson.js:182:15)
remote:     at /home/app/code/lib/configuration.js:440:19
remote:     at Array.forEach (native)
remote:     at Object.<anonymous> (/home/app/code/lib/configuration.js:439:9)
remote:     at Module._compile (module.js:456:26)
remote:     at Object.Module._extensions..js (module.js:474:10)
remote:     at Module.load (module.js:356:32)
remote:     at Function.Module._load (module.js:312:12)
remote:     at Module.require (module.js:364:17)
remote:     at require (module.js:380:17)
remote: ERROR: while postdeploy hook:
remote:
remote: Error: Command failed:
remote: /home/app/code/node_modules/cjson/lib/cjson.js:182
remote:         throw new Error('JSON.parse error - "' + e.message + '"\nFile: "' + pa
remote:               ^
remote: Error: JSON.parse error - "Unexpected token '"
remote: File: "/home/app/code/config/aws.json"
remote:
remote:     at Object.load (/home/app/code/node_modules/cjson/lib/cjson.js:182:15)
remote:     at /home/app/code/lib/configuration.js:440:19
remote:     at Array.forEach (native)
remote:     at Object.<anonymous> (/home/app/code/lib/configuration.js:439:9)
remote:     at Module._compile (module.js:456:26)
remote:     at Object.Module._extensions..js (module.js:474:10)
remote:     at Module.load (module.js:356:32)
remote:     at Function.Module._load (module.js:312:12)
remote:     at Module.require (module.js:364:17)
remote:     at require (module.js:380:17)
remote:
To app@54.220.202.165:git
 + 844bf6f...232b9b1 HEAD -> master (forced update)
```

User error on my part?

**Comment by seanmonstar** [comment]
Sigh. That's likely a typo on my part in aws.json.

**Pull request merged**